app/src/main/java/org/mozilla/focus/bookmark/BookmarkAdapter.java

```diff
@@ -39,43 +34,62 @@ public SiteItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int view
-        holder.rootView.setTag(item.getUuid());
-        holder.textMain.setText(item.getName());
-        holder.textSecondary.setText(item.getAddress());
+        final BookmarkModel item = getItem(position);
+        if (item == null) {
+            return;
+        }
+
+        holder.rootView.setTag(item.getId());
+        holder.textMain.setText(item.getTitle());
+        holder.textSecondary.setText(item.getUrl());
         holder.rootView.setOnClickListener(v -> {
-            ScreenNavigator.get(context).showBrowserScreen(item.getAddress(), true, false);
-            listener.onItemClicked();
-            TelemetryWrapper.bookmarkOpenItem();
+            listener.onItemClicked(item.getUrl());
         });
-        final PopupMenu popupMenu = new PopupMenu(context, holder.btnMore);
-        popupMenu.setOnMenuItemClickListener( menuItem -> {
+        final PopupMenu popupMenu = new PopupMenu(holder.btnMore.getContext(), holder.btnMore);
+        popupMenu.setOnMenuItemClickListener(menuItem -> {
             if (menuItem.getItemId() == R.id.remove) {
-                TempInMemoryBookmarkRepository.getInstance().removeSingle((UUID) holder.rootView.getTag());
-                notifyDataSetChanged();
-                TelemetryWrapper.bookmarkRemoveItem();
+                listener.onItemDeleted(item);
             }
             if (menuItem.getItemId() == R.id.edit) {
-                context.startActivity(new Intent(context, EditBookmarkActivity.class).putExtra(EditBookmarkActivityKt.ITEM_UUID_KEY, item.getUuid()));
-                TelemetryWrapper.bookmarkEditItem();
+                listener.onItemEdited(item);
             }
             return false;
         });
         popupMenu.inflate(R.menu.menu_bookmarks);
-        holder.btnMore.setOnClickListener( v -> {
+        holder.btnMore.setOnClickListener(v -> {
             popupMenu.show();
             TelemetryWrapper.showBookmarkContextMenu();
         });
     }
 
     @Override
     public int getItemCount() {
-        int count = TempInMemoryBookmarkRepository.getInstance().list().size();
-        if (count == 0) {
+        return (bookmarkModels != null ? bookmarkModels.size() : 0);
+    }
+
+    public void setData(List<BookmarkModel> bookmarkModels) {
+        this.bookmarkModels = bookmarkModels;
+        if (getItemCount() == 0) {
             listener.onStatus(PanelFragment.VIEW_TYPE_EMPTY);
         } else {
             listener.onStatus(PanelFragment.VIEW_TYPE_NON_EMPTY);
         }
-        return count;
+        notifyDataSetChanged();
+    }
+
+    private BookmarkModel getItem(int index) {
+        if (index >= 0 && bookmarkModels != null && bookmarkModels.size() > index) {
+            return bookmarkModels.get(index);
+        } else {
+            return null;
+        }
+    }
+
+    public interface BookmarkPanelListener extends PanelFragmentStatusListener {
+        void onItemClicked(String url);
```

Comment:
Maybe rewrite PanelFragmentStatusListener to PanelFragmentStatusListner<I> and void onItemClicked(I item)